### Problem Definition
#### Predict The Purchasing Intentions of a visitor an online store's website

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Load the dataset

data = pd.read_csv(r'C:\Users\vitus.agwulonu\Documents\PERSONAL DOC\Lecture\Internship\online_shoppers_intention.csv')
data.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

In [3]:
cols = len(data.columns[1:])

In [4]:
cols

17

##### Data Inspection and Manipulations

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  int64  
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

In [6]:
data.describe()

Administrative  Administrative_Duration  Informational  \
count    12330.000000             12330.000000   12330.000000   
mean         2.315166                80.818611       0.503569   
std          3.321784               176.779107       1.270156   
min          0.000000                 0.000000       0.000000   
25%          0.000000                 0.000000       0.000000   
50%          1.000000                 7.500000       0.000000   
75%          4.000000                93.256250       0.000000   
max         27.000000              3398.750000      24.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count            12330.000000    12330.000000             12330.000000   
mean                34.472398       31.731468              1194.746220   
std                140.749294       44.475503              1913.669288   
min                  0.000000        0.000000                 0.000000   
25%                  0.000000        7.000000               184.137500   
50%                  0.000000       18.000000               598.936905   
75%                  0.000000       38.000000              1464.157214   
max               2549.375000      705.000000             63973.522230   

        BounceRates     ExitRates    PageValues    SpecialDay  \
count  12330.000000  12330.000000  12330.000000  12330.000000   
mean       0.022191      0.043073      5.889258      0.061427   
std        0.048488      0.048597     18.568437      0.198917   
min        0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.014286      0.000000      0.000000   
50%        0.003112      0.025156      0.000000      0.000000   
75%        0.016813      0.050000      0.000000      0.000000   
max        0.200000      0.200000    361.763742      1.000000   

       OperatingSystems       Browser        Region   TrafficType  
count      12330.000000  12330.000000  12330.000000  12330.000000  
mean           2.124006      2.357097      3.147364      4.069586  
std            0.911325      1.717277      2.401591      4.025169  
min            1.000000      1.000000      1.000000      1.000000  
25%            2.000000      2.000000      1.000000      2.000000  
50%            2.000000      2.000000      3.000000      2.000000  
75%            3.000000      2.000000      4.000000      4.000000  
max            8.000000     13.000000      9.000000     20.000000

In [7]:
# checking for missing values

data.isna().sum()

Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
Month                      0
OperatingSystems           0
Browser                    0
Region                     0
TrafficType                0
VisitorType                0
Weekend                    0
Revenue                    0
dtype: int64

In [8]:
data['Revenue'].value_counts()

False    10422
True      1908
Name: Revenue, dtype: int64

In [9]:
data.shape

(12330, 18)

#### Feature Engineering

In [10]:
data.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

In [11]:
def revenue_column(Revenue):
    if Revenue == True:
        return 1
    else:
        return 0
data['Revenue_category'] = data['Revenue'].apply(revenue_column)

In [12]:
data.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  \
0        1       1            1  Returning_Visitor    False    False   
1        2       1            2  Returning_Visitor    False    False   
2        1       9            3  Returning_Visitor    False    False   
3        2       2            4  Returning_Visitor    False    False   
4        3       1            4  Returning_Visitor     True    False   

   Revenue_category  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [13]:
data.tail()

Administrative  Administrative_Duration  Informational  \
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
12325     0.007143   0.029031   12.241717         0.0   Dec                 4   
12326     0.000000   0.021333    0.000000         0.0   Nov                 3   
12327     0.083333   0.086667    0.000000         0.0   Nov                 3   
12328     0.000000   0.021053    0.000000         0.0   Nov                 2   
12329     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  \
12325        6       1            1  Returning_Visitor     True    False   
12326        2       1            8  Returning_Visitor     True    False   
12327        2       1           13  Returning_Visitor     True    False   
12328        2       3           11  Returning_Visitor    False    False   
12329        2       1            2        New_Visitor     True    False   

       Revenue_category  
12325                 0  
12326                 0  
12327                 0  
12328                 0  
12329                 0

In [14]:
data['Revenue_category'].value_counts()

0    10422
1     1908
Name: Revenue_category, dtype: int64

In [15]:
#separated target from dataset
y = data["Revenue_category"]
X = data.drop("Revenue_category", axis=1)

In [16]:
X

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.200000   0.200000    0.000000         0.0   Feb                 1   
1         0.000000   0.100000    0.000000         0.0   Feb                 2   
2         0.200000   0.200000    0.000000         0.0   Feb                 4   
3         0.050000   0.140000    0.000000         0.0   Feb                 3   
4         0.020000   0.050000    0.000000         0.0   Feb                 3   
...            ...        ...         ...         ...   ...               ...   
12325     0.007143   0.029031   12.241717         0.0   Dec                 4   
12326     0.000000   0.021333    0.000000         0.0   Nov                 3   
12327     0.083333   0.086667    0.000000         0.0   Nov                 3   
12328     0.000000   0.021053    0.000000         0.0   Nov                 2   
12329     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0            1       1            1  Returning_Visitor    False    False  
1            2       1            2  Returning_Visitor    False    False  
2            1       9            3  Returning_Visitor    False    False  
3            2       2            4  Returning_Visitor    False    False  
4            3       1            4  Returning_Visitor     True    False  
...        ...     ...          ...                ...      ...      ...  
12325        6       1            1  Returning_Visitor     True    False  
12326        2       1            8  Returning_Visitor     True    False  
12327        2       1           13  Returning_Visitor     True    False  
12328        2       3           11  Returning_Visitor    False    False  
12329        2       1            2        New_Visitor     True    False  

[12330 rows x 18 columns]

In [17]:
X = data.drop(columns=["Revenue", "Revenue_category"], axis=1)

In [18]:
X

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.200000   0.200000    0.000000         0.0   Feb                 1   
1         0.000000   0.100000    0.000000         0.0   Feb                 2   
2         0.200000   0.200000    0.000000         0.0   Feb                 4   
3         0.050000   0.140000    0.000000         0.0   Feb                 3   
4         0.020000   0.050000    0.000000         0.0   Feb                 3   
...            ...        ...         ...         ...   ...               ...   
12325     0.007143   0.029031   12.241717         0.0   Dec                 4   
12326     0.000000   0.021333    0.000000         0.0   Nov                 3   
12327     0.083333   0.086667    0.000000         0.0   Nov                 3   
12328     0.000000   0.021053    0.000000         0.0   Nov                 2   
12329     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  
0            1       1            1  Returning_Visitor    False  
1            2       1            2  Returning_Visitor    False  
2            1       9            3  Returning_Visitor    False  
3            2       2            4  Returning_Visitor    False  
4            3       1            4  Returning_Visitor     True  
...        ...     ...          ...                ...      ...  
12325        6       1            1  Returning_Visitor     True  
12326        2       1            8  Returning_Visitor     True  
12327        2       1           13  Returning_Visitor     True  
12328        2       3           11  Returning_Visitor    False  
12329        2       1            2        New_Visitor     True  

[12330 rows x 17 columns]

In [19]:
# Encoding categorical variables

X = pd.get_dummies(X)
X

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay  ...  Month_Jul  \
0         0.200000   0.200000    0.000000         0.0  ...          0   
1         0.000000   0.100000    0.000000         0.0  ...          0   
2         0.200000   0.200000    0.000000         0.0  ...          0   
3         0.050000   0.140000    0.000000         0.0  ...          0   
4         0.020000   0.050000    0.000000         0.0  ...          0   
...            ...        ...         ...         ...  ...        ...   
12325     0.007143   0.029031   12.241717         0.0  ...          0   
12326     0.000000   0.021333    0.000000         0.0  ...          0   
12327     0.083333   0.086667    0.000000         0.0  ...          0   
12328     0.000000   0.021053    0.000000         0.0  ...          0   
12329     0.000000   0.066667    0.000000         0.0  ...          0   

       Month_June  Month_Mar  Month_May  Month_Nov  Month_Oct  Month_Sep  \
0               0          0          0          0          0          0   
1               0          0          0          0          0          0   
2               0          0          0          0          0          0   
3               0          0          0          0          0          0   
4               0          0          0          0          0          0   
...           ...        ...        ...        ...        ...        ...   
12325           0          0          0          0          0          0   
12326           0          0          0          1          0          0   
12327           0          0          0          1          0          0   
12328           0          0          0          1          0          0   
12329           0          0          0          1          0          0   

       VisitorType_New_Visitor  VisitorType_Other  \
0                            0                  0   
1                            0                  0   
2                            0                  0   
3                            0                  0   
4                            0                  0   
...                        ...                ...   
12325                        0                  0   
12326                        0                  0   
12327                        0                  0   
12328            

In [20]:
y

0        0
1        0
2        0
3        0
4        0
        ..
12325    0
12326    0
12327    0
12328    0
12329    0
Name: Revenue_category, Length: 12330, dtype: int64

#### Model Selection, Training, and Validation

In [21]:
# Split the data into training and testing sets

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=0)

In [22]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(9864, 28)
(2466, 28)
(9864,)
(2466,)


#### Train and test 2 supervised Model

In [44]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB

In [41]:
# Logistic Regression
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)
log_reg_pred = log_reg.predict(X_test)

C:\Users\vitus.agwulonu\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [25]:
print(log_reg_pred[:5])

[0 0 0 0 0]


In [26]:
rf = RandomForestClassifier(n_estimators=100)
rf.fit(X_train, y_train)
rf_pred = rf.predict(X_test)

In [27]:
print(rf_pred[:5])

[0 0 0 0 0]


In [45]:
nb_clf = GaussianNB()
nb_clf.fit(X_train, y_train)
nb_clf_pred = nb_clf.predict(X_test)

In [37]:
from sklearn.metrics import classification_report

In [47]:
# Classification report for Logistic Regression
print(classification_report(y_test, log_reg_pred))
print()

# Classification report for RandomForestClassifier
print(classification_report(y_test, rf_pred))
print()

# Classification report for Gauaaian Classifier
print(classification_report(y_test, nb_clf_pred))
print()

              precision    recall  f1-score   support

           0       0.88      0.97      0.92      2044
           1       0.74      0.36      0.48       422

    accuracy                           0.87      2466
   macro avg       0.81      0.66      0.70      2466
weighted avg       0.86      0.87      0.85      2466


              precision    recall  f1-score   support

           0       0.91      0.96      0.94      2044
           1       0.76      0.55      0.64       422

    accuracy                           0.89      2466
   macro avg       0.84      0.76      0.79      2466
weighted avg       0.89      0.89      0.89      2466


              precision    recall  f1-score   support

           0       0.91      0.83      0.87      2044
           1       0.42      0.61      0.50       422

    accuracy                           0.79      2466
   macro avg       0.67      0.72      0.68      2466
weighted avg       0.83      0.79      0.80      2466




In [48]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, log_reg_pred))
print()
print()
print(confusion_matrix(y_test, rf_pred))
print()
print()
print(confusion_matrix(y_test, nb_clf_pred))

[[1991   53]
 [ 272  150]]


[[1969   75]
 [ 188  234]]


[[1690  354]
 [ 164  258]]


In [33]:
from sklearn.metrics import accuracy_score

In [35]:
accuracy = accuracy_score(y_test, log_reg_pred)
print("Accuracy: {:.2f}".format(accuracy))

Accuracy: 0.87


In [36]:
accuracy = accuracy_score(y_test, rf_pred)
print("Accuracy: {:.2f}".format(accuracy))

Accuracy: 0.89


In [46]:
accuracy = accuracy_score(y_test, nb_clf_pred)
print("Accuracy: {:.2f}".format(accuracy))

Accuracy: 0.79


### Model/Result Interpretation

#### The classification report presents the performance evaluation metrics of three different classification models: 
#### - Logistic Regression, GaussianNB , and Random Forest Classifier (RF).

#### - Each report shows the precision, recall, and F1-score for each class, as well as the support and accuracy for the model.

#### - Precision is the ratio of correctly predicted positive observations to the total predicted positive observations, 
#### -  while recall is the ratio of correctly predicted positive observations to the total actual positive observations. 
#### - F1-score is the harmonic mean of precision and recall.
#### - The support refers to the number of observations in each class, while the accuracy is the overall percentage of correctly classified observations.

#### Precision = TP/(TP+FP)
#### Recall = TP/(TP+FN)
#### F1 Score = 2*(Precision * Recall)/(Precision + Recall)
#### Accuracy = (TP+TN)/Total
#### Where; 
##### TN = True Negative, TP = True Positive, FN = False Negative, FP = False Positive

#### 5a. From my result, the Precision rate for Logistic Regression is 74% and its Recall rate is 36%, resulting in an F1-score of 48%. However, due to the relatively low Recall rate, Logistic Regression may not be as effective in predicting accurately the purchasing intentions of online visitors to the websit compared to other classification models.


#### The Random Forest Classifier (RF) is a reliable tool for predicting the purchasing intentions of online visitors to a website due to its impressive Recall rate of 55%. Nonetheless, the model's high Precision rate of 76% could potentially cause inability of the model to predict accurately the purchasing intentions of online visitors.

#### GaussianNB Classifier - Based on the results, the GaussianNB Classifier appears to be the most effective out of the three classifiers that were utilized. This particular model achieved a high Recall rate of 61% for predicting positive purchasing intentions of online visitors to the website, which was the primary objective of training and testing the models.


#### 5b. What Matrics are the most important:
##### The most important Matric is the Recal Matric as it is able to predict positve purchasing intentions of online visitors in the dataset. A higher recall indicates that the model has fewer false negatives, which means the model is better at identifying all the positive instances. Hence the business should be more concerned about the false negatives.

#### https://github.com/Agwulonu/Online-Shoppers-Purchasing-Intentions